gap/attr.gi

```diff
@@ -128,18 +135,29 @@ function(digraph)
     if not IsNullDigraph(subdigraph)
         and not IsBipartiteDigraph(subdigraph) then
       if IsCompleteDigraph(DigraphSymmetricClosure(subdigraph)) then
-        if n > chrom then
+        if chrom = fail or n > chrom then
           chrom := n;
         fi;
       else
-        for i in [3 .. n - 1] do
-          if DigraphColoring(subdigraph, i) <> fail then
-            if i > chrom then
-              chrom := i;
+        upper := RankOfTransformation(DigraphColouring(subdigraph), nr);
+        if chrom = fail or upper > chrom then
+          upper := Minimum(upper, nr - 1);
+          lower := Maximum(chrom, CliqueNumber(subdigraph));
+          if upper = lower then
+            # This line is hard to test since DigraphColouring is
+            # non-deterministic now
+            chrom := upper; # upper > chrom
+          else # lower < upper
+            chrom := upper - 1;
+            repeat
+              coloring := DigraphColouring(subdigraph, chrom);
+              chrom    := chrom - 1;
+            until coloring = fail or chrom = lower;
```

Comment:
I get the `until coloring = fail` bit. If `coloring = fail`, then the connected component can not be coloured with `chrom + 1` colours, but since we know that we can colour it with `chrom + 2` colours, `chrom + 2` must be the chromatic number of this connected component. And hence `chrom + 2` is how many colours we need to colour the whole graph, so far. Great.

I'm unsure about the other condition. If `chrom = lower`, then we know that we can colour the connected component with `lower + 1` colours, and that the connected component is allowed to use, or requires, `lower` colours. However, I think it's possible that we never actually test whether this connected component *can* in fact be coloured with `lower` colours - it might not be possible to do it with exactly `lower` colours. The logic of the current code seems to suggest that if the loop terminates because `chrom = lower`, then the connected component *can* be coloured using only `lower` colours. However I'm not sure that this is guaranteed.

I'll try and find an example.